src/main/java/com/puppycrawl/tools/checkstyle/Main.java

```diff
@@ -259,613 +163,588 @@ public static void main(String... args) throws IOException {
+                    result.add("Option '-s' cannot be used with other options.");
+                }
+                else if (filesToProcess.size() > 1) {
+                    result.add("Printing xpath suppressions is allowed for only one file.");
+                }
+            }
+            else if (hasConfigurationFile) {
+                try {
+                    // test location only
+                    CommonUtil.getUriByFilename(configurationFile);
+                }
+                catch (CheckstyleException ignored) {
+                    final String msg = "Could not find config XML file '%s'.";
+                    result.add(String.format(Locale.ROOT, msg, configurationFile));
+                }
 
-        final Configuration[] children = config.getChildren();
-        for (Configuration child : children) {
-            if ("TreeWalker".equals(child.getName())) {
-                result = child;
-                break;
+                // validate optional parameters
+                if (propertiesFile != null && !propertiesFile.exists()) {
+                    result.add(String.format(Locale.ROOT,
+                            "Could not find file '%s'.", propertiesFile));
+                }
+                if (checkerThreadsNumber < 1) {
+                    result.add("Checker threads number must be greater than zero");
+                }
+                if (treeWalkerThreadsNumber < 1) {
+                    result.add("TreeWalker threads number must be greater than zero");
+                }
+            }
+            else {
+                result.add("Must specify a config XML file.");
             }
-        }
-        return result;
-    }
 
-    /**
-     * Creates a new instance of the root module that will control and run
-     * Checkstyle.
-     * @param name The name of the module. This will either be a short name that
-     *        will have to be found or the complete package name.
-     * @param moduleClassLoader Class loader used to load the root module.
-     * @return The new instance of the root module.
-     * @throws CheckstyleException if no module can be instantiated from name
-     */
-    private static RootModule getRootModule(String name, ClassLoader moduleClassLoader)
-            throws CheckstyleException {
-        final ModuleFactory factory = new PackageObjectFactory(
-                Checker.class.getPackage().getName(), moduleClassLoader);
+            return result;
+        }
+
+        /**
+         * Do execution of CheckStyle based on Command line options.
+         * @param filesToProcess the list of files whose style to check
+         * @return number of violations
+         * @throws IOException if a file could not be read.
+         * @throws CheckstyleException if something happens processing the files.
+         */
+        private int runCli(List<File> filesToProcess) throws IOException, CheckstyleException {
```

Comment:
Why CliOption now it in change to run Checkstyle ?
Cli is just one way to configure execution, and it collects all parameters, and this class should only do CLI related logic and nothing else. All orchestration of how to run checkstyle should stay in Main class.

Reply:
I see your point that it is better to focus on the responsibilities of each class.

I just submitted a new PR (now going through CI) where responsibilities are divided like this:

**CliOptions responsibilities**

* defines the command line options and usage help message descriptions
* validating the user input
* listing the target files (after figuring out regex patterns, exclusions, etc.)

**Main responsibilities** 

* delegate some work to CliOptions
* run checkstyle


![checkstyle](https://user-images.githubusercontent.com/5917763/47273096-d3999280-d5c9-11e8-8c36-466cf35160b2.png)